src/R8R.sol

```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {ReentrancyGuard} from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {IERC20} from "@openzeppelin/contracts/interfaces/IERC20.sol";
+
+contract R8R is Ownable, ReentrancyGuard {
+    // =========================
+    // === STORAGE VARIABLES ===
+    // =========================
+
+    address public robot;
+    uint256 public gameId;
+    uint256 public gameEntryPriceInEth;
+
+    mapping(uint256 gameId => Game) games;
+
+    IERC20[] public gameTokensAllowedList; // list of ERC20 tokens users can use to enter the game
+    uint256[] public gameTokensEntryPriceList; // list of entry price for the ERC20 tokens
+
+    struct Game {
+        uint256 gameId;
+        address[] players;
+        uint256[] playerRatings;
+        uint256 robotRating;
+        address winner;
+        uint256 winnerPayout;
+    }
+
+    // ==============
+    // === EVENTS ===
+    // ==============
+
+    event GameCreated(uint256 gameId);
+    event PlayerJoinedGameWithEth(address player, uint256 gameId);
+    event PlayerJoinedGameWithTokens(address player, uint256 gameId);
+    event GameEnded();
+
+    // ==============
+    // === ERRORS ===
+    // ==============
+    error Error__TokenTransferFailed();
+
+    // ===================
+    // === CONSTRUCTOR ===
+    // ===================
+
+    constructor(
+        address _robot,
+        uint256 _gameEntryPriceInEth,
+        IERC20[] memory _gameTokensAllowedList,
+        uint256[] memory _gameTokensEntryPriceList
+    ) Ownable(msg.sender) {
+        robot = _robot;
+        gameEntryPriceInEth = _gameEntryPriceInEth;
+        gameTokensAllowedList = _gameTokensAllowedList;
+        gameTokensEntryPriceList = _gameTokensEntryPriceList;
+    }
+
+    // =========================
+    // === PUBLIC FUNCTIONS ====
+    // =========================
+
+    // @notice called by AI wallet to initialise a new game
+    // @notice creates a new Game struct that players + ratings can then be added to when they join
+    function createGame() public onlyRobot {
+        // increment gameId
+        gameId = gameId + 1;
+
+        // empty placeholder arrays for players + their ratings (until someone joins the game)
+        address[] memory playersArray;
+        uint256[] memory playersRatingsArray;
```

Comment:
Could we turn these two variables into a mapping and just map address to player rating?